### ISSUE TITLE
`AllIndices` type for identity indexing

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaLang/julia#24069]
With the proposal for https://github.com/JuliaLang/julia/issues/22858, I've been thinking it could be useful to have a type which commonly represents "AllIndices". So `A[AllIndices()]` would return `A`, making it the identity indexing element in some sense.

The reason is so that this could handle non-standard indexing. In some algorithms you might want the user to specify some indices that they "want to save", and you can default to `AllIndices`. In many of these cases, you haven't built the array yet and may be getting a non-standard type from the user, so defaulting to `eachindex` or something like that isn't possible. Currently a way to handle this is to default the chosen indices to `nothing`, and if you see `nothing` you treat it as `AllIndices` applied to the type by iterating `eachindex` after the array is built.

The name could change of course.

**Comment by mbauman** [comment]
How is this different from `A[:]`?

**Comment by ChrisRackauckas** [comment]
`Colon` doesn't keep shape:

```julia
julia> A = rand(2,2)
2×2 Array{Float64,2}:
 0.138581  0.120538
 0.528131  0.11805

julia> A[:]
4-element Array{Float64,1}:
 0.138581
 0.528131
 0.120538
 0.11805
```

**Comment by mbauman** [comment]
This could be solved with #5405 as a base case (just the ellipsis).

**Comment by mbauman** [comment]
Now this is a duplicate of #5405.  A single ellipsis represents this operation well, and it's now in EllipsisNotation.jl as `A[..]`.
